### PR TITLE
[JENKINS-55050] Reduce performance impact of SEC-904 (esp. zip)

### PR DESCRIPTION
**Pull request by Wadeck**
Alternative to #3848 (advanced version).

See [JENKINS-55050](https://issues.jenkins-ci.org/browse/JENKINS-55050).

- [Google spreadsheet](https://docs.google.com/spreadsheets/d/1uHtAZvd0TNWdFFeR6D7BwZdq3cJ22UHOZrpV1q32o_E) to explain the performance improvement

There are multiples methods that are impacted depending on the scenario. 

#### Scenario 1), workspace files located on master

**Initial situation:** The `FilePathVF#list(...)` with Glob was used. This means it list all the file recursively using Ant library. _This invocation is done on the master directly._
**SEC-904:** The correction added a check for every files if they are descendant of the root (= workspace / artifact base folder). _The check was done on the master._

This PR proposes to replace this method by recursive listing of file (see `VirtualFile.FilePathVF#listOnlyDescendants` and `DirectoryBrowserSupport#collectRecursivelyAllLegalChildren()`), taking into consideration only the **descendants** of the root. As we are doing the check directly on a child, we can optimize the call to avoid most of the time, to go back to the root. In addition, when we reach a symlink to outside of the root, we do not explore it, compared to the initial version, leading to useless computation. _The checks is still done on the master._

#### Scenario 2), workspace files located on agent

**Initial situation:** The `FilePathVF#list(...)` with Glob was used. This means it list all the file recursively using Ant library. _This invocation is done on the agent._
**SEC-904:** The correction added a check for every files if they are descendant of the root (= workspace / artifact base folder). That check was done on the agent, with a new remoting call every time.

This PR proposes an improve method in Glob, that will check the file, directly inside the agent invocation (without going back to master) with a caching strategy. The use of Glob cannot be interrupted, in the sense, we cannot stop the search at a given folder because we know it's outside of the root folder. Using the caching, we reduce drastically the number of real check that require to be done. That concern especially `DirScanner.DescendantFileSelector`. We distinguish the two first scenario by checking the channel, so that we can decide on which technique we want to apply (distinction done using `VirtualFile#supportQuickRecursiveListing()`). _Those checks are done in the agent directly, without further communication requirement._
Without that distinction, the performance was decreased by an huge amount if we used the recursive listing on agent due to the additional number of calls.

#### Scenario 3), artifacts files (located on master, not possible to have them on agent)

**Initial situation:** The `FileVF#list(...)` with Glob was used. This means it list all the file recursively using Ant library. _This invocation is done on the master._
**SEC-904:** The correction added a check for every files if they are descendant of the root (= workspace / artifact base folder). Those checks were done on the master.

For this scenario, it's quite the same as for the scenario 1), but using `VirtualFile.FileFV#listOnlyDescendants()`. Using the `File` API, we can directly list only the descendant files and create the related `VirtualFile` with a prefilled information about their descendancy. That reduce the number of check that were done. For a single `zipEntry`, the previous code was checking the descendency **thrice**, now the information is cached. The cache is strictly bounded to the request, without any side effect.

#### Summary of performance gain

- cut in half the time required for scenario 1.
- 10% gain for scenario 2. (without the distinction, the performances were slowed down by 50% after the initial optimization)
- Up to 10x faster for scenario 3 (that was the goal of the initial ticket).

### Proposed changelog entries

* Reduce the impact on the performance from the SECURITY-904 correction. That concerns especially the `all files in zip` feature proposed by Workspace and Artifact views in a job.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck

### CODE REVIEW

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (jvz) [inline core/src/main/java/jenkins/util/VirtualFile.java]:
```suggestion
            if (potentialChildRelativePath.isEmpty() && cacheDescendant) {
```

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (jvz) [inline core/src/main/java/jenkins/util/VirtualFile.java]:
```suggestion
                if (isDescendant && potentialChildRelativePath.isEmpty()) {
```

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: I added a `return` as we do not want to continue actually.

File: core/src/main/java/hudson/FilePath.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/FilePath.java]: add system logging then? Otherwise the issue is just suppressed

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: add it to `finally`? Otherwise ZIP stream may be corrupted if `cf.open()` or copy fail

Comment (Wadeck) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Good improvement to the existing code

File: core/src/main/java/hudson/FilePath.java
Comment (Wadeck) [reply core/src/main/java/hudson/FilePath.java]: Warning added with the exception.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/VirtualFile.java]: dead code

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/VirtualFile.java]: NIT: support**s**QuickRecursiveListing

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/VirtualFile.java]: pre-initialize by `children.length` ?

**Review by oleg-nenashev — COMMENTED** [review]
I do not see anything wrong here though the code becomes much more complex to get a performance improvement. As expected, I'd guess

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: done for the two occurences

**Comment by oleg-nenashev** [comment]
@Wadeck merge conflict

**Review by jeffret-b — APPROVED** [review]
Looks like it should work. As Oleg notes, it would be nice if the solution were simpler, but I don't see any way to achieve that and also the performance gains. And in some of the scenarios this still leaves a significant degradation from what used to be. Again, unfortunately I don't have any better ideas.

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (varyvol) [inline core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Why not join declaration and assignment?

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (varyvol) [inline core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Is the replace necessary? Both the map and the list of names seem to generate the relative paths with '/' already.

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (varyvol) [inline core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Why not using te implicit for loop?

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (varyvol) [inline core/src/main/java/jenkins/util/VirtualFile.java]: What's the difference between this and listOnlyDescendants then?

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (varyvol) [inline core/src/main/java/jenkins/util/VirtualFile.java]: Is this class used?

File: core/src/main/java/hudson/util/DirScanner.java
Comment (jvz) [inline core/src/main/java/hudson/util/DirScanner.java]: These fields can be `final` it looks like.

File: core/src/main/java/hudson/util/DirScanner.java
Comment (jvz) [inline core/src/main/java/hudson/util/DirScanner.java]: Isn't the file separator always length 1? You can access it via `File.separatorChar` which is length 1.

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (vilacides) [inline core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Is this todo still applicable?

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (jvz) [inline core/src/main/java/jenkins/util/VirtualFile.java]: Did you mean:

```suggestion
                return f.act(new Scanner2(includes, excludes, useDefaultExcludes));
```

**Review by jvz — CHANGES_REQUESTED** [review]
Got a couple small things to look at, otherwise looks good.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (vilacides) [inline core/src/main/java/jenkins/util/VirtualFile.java]: this code is duplicated, shouldn't we encapsulate it for maintainability's sake?

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (Wadeck) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Legacy code just moved. I preferred to not touch the code especially with the TODO comment above.

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (Wadeck) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Changed! good point, the code like this was more useful in the legacy version

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (Wadeck) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: Good point, too many iterations of that code :D

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: > Lists only the children that are descendant of the root directory

The important concept is the "root". It's not the current file, it's the file that generates directly or indirectly that current file.

In our case, the root is the workspace or the artifact folder, the current VirtualFile that will call this method, could be 2 layers "inside" that root folder. The `list` will list all the children of the current folder, but the `listOnlyDescandants` will only list the children that are effectively also "inside" the root folder (forbiding the symlink to outside of the workspace).

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (Wadeck) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: @jglick Do you remember why you put the TODO there ? 

The initial code used the `.list(...)` method that was already a callable, but the stream comes from the response. Did you think about sending the zip from the callable directly ?

File: core/src/main/java/hudson/util/DirScanner.java
Comment (Wadeck) [reply core/src/main/java/hudson/util/DirScanner.java]: Yeah they could, as there was no external usage, I was wondering. But you're right, I change it!

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: No it was just during my testing. I changed the behavior of the Scanner that now consider only the descendants of the root. I thought initially I will need to split the behavior in two but that was not the case in "final" version.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (varyvol) [reply core/src/main/java/jenkins/util/VirtualFile.java]: Sorry I didn't explain myself clearly. I understand what you just explained, but my comment was because in the javadoc for the "list" method you added the "one level deep" comment, which is not true for such method IIUC.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: Good point, I put the method in the VirtualFile class directly. That will also remove the duplication of comment.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: > which is not true for such method IIUC

Could you elaborate ?

Initially I added the "one level deep" because I investigated the `list()` and `list(...)`, the version with arguments is actually recursive, compared to the no-arg version. I was confused by this naming, so I added the comment.

**Review by jvz — APPROVED** [review]

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (jglick) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: No, `run(Callable)` is about batch computation. Not to be confused with `hudson.remoting.Callable`!

See [JEP-202](https://jenkins.io/jep/202). Currently when you download a `*.zip` URL containing a number of artifacts (for example the Incrementals publisher function does this), some `VirtualFile` methods (`list`, `open`, `lastModified`) will be called one by one, asymptotically proportional to the number of files. For artifacts stored in `$JENKINS_HOME`, this is fine, but if you are using `artifact-manager-s3` it would be rather inefficient. Using `VirtualFile.run(Callable)` here would allow for a faster implementation—in the case of S3, one metadata API call plus one blob download per artifact.

(Even better would be a URL pattern returning some text response like

```
target/stuff-1.0.pom https://s3.whatever…/…/target/stuff-1.0.pom?authentication=…
target/stuff-1.0.jar https://s3.whatever…/…/target/stuff-1.0.jar?authentication=…
```

so that the master could avoid downloading blobs at all.)

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (varyvol) [reply core/src/main/java/jenkins/util/VirtualFile.java]: So even for the list() method, it is only one level deep?

Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: The `list()` and `listOnlyDescandants()` are ony one level deep but not the `list(...)`

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (Wadeck) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: @jglick thanks for the precision, really appreciated. Are you OK if we let the TODO in place for future improvement ?

Comment (jglick) [reply core/src/main/java/hudson/model/DirectoryBrowserSupport.java]: If you are not intending to introduce a `run(Callable)` call then the comment should be left alone.

**Comment by oleg-nenashev** [comment]
@jglick are you satisfied by the answer?

**Comment by oleg-nenashev** [comment]
I will not be merging it to the next weekly, because it seems there are some outstanding comments

**Comment by Wadeck** [comment]
> [...] can we cheaply & simply just ignore _all_ symlinks [...]

@jglick I created [JENKINS-55970](https://issues.jenkins-ci.org/browse/JENKINS-55970) for that task. We cannot propose such feature removal while letting the current status not really good in terms of performance. From my PoV the perf improvement is quickly approved, compared to the removal of the symlink support in workspace/archive that could take a while to be accepted.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/util/VirtualFile.java]: It will suppress the interruption logic for the thread. I do not think it is good to suppress that, and I would rather prefer to see it propagates as far as API compatibility requirements allow

File: core/src/main/java/hudson/util/DirScanner.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/DirScanner.java]: Should not be interrupted

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Thanks @Wadeck !

It significantly increases the code complexity, but I do not see errors here except the InterruptedException suppression. It is dangerous for thread management and safe termination in Java, and such exceptions should not be suppressed when it is possible

**Comment by jglick** [comment]
> the perf improvement is quickly approved, compared to the removal of the symlink support in workspace/archive that could take a while to be accepted

:man_shrugging: so long as we are not introducing any public APIs which would possibly be rendered obsolete by removal of symlink traversal, I see no harm in it.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/VirtualFile.java]: can be handled properly by extending API of the restricted method

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/util/VirtualFile.java]: My comment is still actual here

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: @oleg-nenashev I do not see any reason to do that when the rest of the methods "just" wrap it within `IOException`, the behavior will not be consistent.

File: core/src/main/java/hudson/util/DirScanner.java
Comment (Wadeck) [reply core/src/main/java/hudson/util/DirScanner.java]: @oleg-nenashev What do you propose here? Wrapping it inside a `RuntimeException`, throwing an `AssertException` ? The method signature does not allow the throw of InterruptedException here.

File: core/src/main/java/jenkins/util/VirtualFile.java
Comment (Wadeck) [reply core/src/main/java/jenkins/util/VirtualFile.java]: (I saw the conversation marked as resolved but not sure who did that)
It's considered in the next commit, wrapping it inside an `IOException`.

**Review by oleg-nenashev — APPROVED** [review]
@Wadeck I have no bandwidth to follow-up on my comments and to show examples right now. `InterruptedException` is often suppressed in Jenkins and in the classes touched by this PR, so it is a wide issue. Although I still have concerns about it, I do not want to block this PR taking its importance

I have created https://issues.jenkins-ci.org/browse/JENKINS-56041to follow-up on my comments. Will merge once the build passes

**Pull request merged**
